odoo/tools/misc.py

```diff
@@ -135,124 +135,66 @@ def exec_pg_command_pipe(name, *args):
+    raise IOError("File not found: " + file_path)
 
+def file_open(name, mode="r", filter_ext=None):
+    """Open a file from within the addons_path directories, as an absolute or relative path.
 
-def _fileopen(path, mode, basedir, pathinfo, basename=None):
-    name = os.path.normpath(os.path.normcase(os.path.join(basedir, path)))
+    Example::
 
-    paths = odoo.addons.__path__ + [config['root_path']]
-    for addons_path in paths:
-        addons_path = os.path.normpath(os.path.normcase(addons_path)) + os.sep
-        if name.startswith(addons_path):
-            break
-    else:
-        raise ValueError("Unknown path: %s" % name)
-
-    if basename is None:
-        basename = name
-    # Give higher priority to module directories, which is
-    # a more common case than zipped modules.
-    if os.path.isfile(name):
-        if 'b' in mode:
-            fo = open(name, mode)
-        else:
-            fo = io.open(name, mode, encoding='utf-8')
-        if pathinfo:
-            return fo, name
-        return fo
-
-    # Support for loading modules in zipped form.
-    # This will not work for zipped modules that are sitting
-    # outside of known addons paths.
-    head = os.path.normpath(path)
-    zipname = False
-    while os.sep in head:
-        head, tail = os.path.split(head)
-        if not tail:
-            break
-        if zipname:
-            zipname = os.path.join(tail, zipname)
-        else:
-            zipname = tail
-        zpath = os.path.join(basedir, head + '.zip')
-        if zipfile.is_zipfile(zpath):
-            zfile = zipfile.ZipFile(zpath)
-            try:
-                fo = io.BytesIO()
-                fo.write(zfile.read(os.path.join(
-                    os.path.basename(head), zipname).replace(
-                        os.sep, '/')))
-                fo.seek(0)
-                if pathinfo:
-                    return fo, name
-                return fo
-            except Exception:
-                pass
-    # Not found
-    if name.endswith('.rml'):
-        raise IOError('Report %r does not exist or has been deleted' % basename)
-    raise IOError('File not found: %s' % basename)
+    >>> file_open('hr/static/description/icon.png')
+    >>> with file_open('/opt/odoo/addons/hr/static/description/icon.png', 'rb') as f:
+    >>>     f.read()
 
+    :param name: absolute or relative path to a file located inside an addon
+    :param mode: file open mode, as for `open()`
+    :param list(str) filter_ext: optional list of supported extensions (without leading dot)
+    :return: file object, as returned by `open()`
+    :raise IOError: if the file is not found under the known `addons_path` directories
+    :raise ValueError: if the file doesn't have one of the supported extensions (`filter_ext`)
+    """
+    path = file_path(name, filter_ext=filter_ext)
+    if os.path.isfile(path):
+        return open(path, mode)
```

Comment:
Needs the `encoding` if not opening in binary mode, otherwise encoding will be selected based on `locale.getpreferredencoding(do_setlocale=False)` which can be any random garbage set in LC_whatever. Sadly `open` does not allow (and ignore) `encoding` in case of a binary file open, so it needs to be a conditional path as with the original.

Reply:
For the record, managed to reproduce this by installing + forcing a non-utf8 and non-C locale on Python 3.8

```sh
$ sudo locale-gen fr_FR
Generating locales (this might take a while)...
  fr_FR.ISO-8859-1... done
Generation complete.
$ LC_ALL=fr_FR.iso8859-1 python3 -c 'import locale; print(locale.getpreferredencoding(False))'
ISO-8859-1
```

Python 3.7 comes with PEP-538 that modifies the semantics of the fallback `C` and `POSIX` to turn them into `C.utf-8` instead of the default `ascii` version. This makes the problem less likely to occur, but we never know what funky locales people are going to use.

Python 3.7 also comes with PEP-540 so we could use `-X utf8` or `PYTHONUTF8=1` to force the new utf8 mode, unfortunately this cannot be done from inside the Python process (the locale is already set, and `sys.flags.ut8_mode` is readonly). As we can't reliably modify the invocation parameters of the Odoo server (people run it in many different ways), we won't be able to count on that.

So we'll keep the forced encoding for text mode `file_open()`, and since it's very hard to cover that with a unit test due to the global and system-dependent locale stuff, we'll just keep a good comment ;-)